src/dstack/_internal/core/backends/oci/resources.py

```diff
@@ -502,3 +539,89 @@ def update_route_table(
             ]
         ),
     ).data
+
+
+def get_or_create_security_group(
+    name: str, vcn_id: str, compartment_id: str, client: oci.core.VirtualNetworkClient
+) -> oci.core.models.NetworkSecurityGroup:
+    if security_groups := client.list_network_security_groups(
+        display_name=name, vcn_id=vcn_id, compartment_id=compartment_id
+    ).data:
+        return security_groups[0]
+
+    return client.create_network_security_group(
+        oci.core.models.CreateNetworkSecurityGroupDetails(
+            display_name=name, vcn_id=vcn_id, compartment_id=compartment_id
+        )
+    ).data
+
+
+def update_security_group_rules_for_runner_instances(
+    security_group_id: str, client: oci.core.VirtualNetworkClient
+) -> None:
+    rules = [
+        SecurityRule(
+            description="Allow all traffic within this security group",
+            direction=oci.core.models.AddSecurityRuleDetails.DIRECTION_INGRESS,
+            source_type=oci.core.models.AddSecurityRuleDetails.SOURCE_TYPE_NETWORK_SECURITY_GROUP,
+            source=security_group_id,
+            protocol="all",
+        ),
+    ]
+    update_security_group_rules(security_group_id, rules, client)
```

Comment:
What allows ingress traffic on port 22? Is it allowed by default in OCI? Won't it be deleted when we remove all existing security groups? Please leave a comment if it's not obvious.

Reply:
Yes, SSH is allowed by subnet's default security list. Security lists are separate from security groups, actions on security groups do not affect the security list